cmd/upload.go

```diff
@@ -13,30 +13,248 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+of the release artifact and uploads it to the rekor server.`,
 	Run: func(cmd *cobra.Command, args []string) {
 		log := log.Logger
-		log.Info("Upload Called")
-		// rekorServer := viper.GetString("rekor_server")
-		// url := rekorServer + "/api/v1/upload"
+		rekorServer := viper.GetString("rekor_server")
+		url := rekorServer + "/api/v1/add"
+		signature := viper.GetString("signature")
+		publicKey := viper.GetString("public-key")
+		artifactURL := viper.GetString("artifact-url")
+
+		// Before we download anything or validate the signing
+		// Let's check the formatting is correct, if not we
+		// exit and allow the user to resolve their corrupted
+		// GPG files.
+		sig, err := app.FormatSignature(signature)
+		if err != nil {
+			log.Error("Signature validation failed: ", err)
+			os.Exit(1)
+		}
+
+		pub_key, err := app.FormatPubKey(publicKey)
+		if err != nil {
+			log.Error("Pubic key validation failed: ", err)
+			os.Exit(1)
+		}
+
+		// Download the artifact set within flag artifactURL
+
+		log.Info("Download artifact..")
+
+		resp, err := http.DefaultClient.Get(artifactURL)
+		if err != nil {
+			log.Error(err)
+		}
+
+		defer resp.Body.Close()
+
+		log.Info("Contents fetched..")
+
+		readBody, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			log.Error("Error reading response body: ", err)
+		}
+
+		// Generate Hash for downloaded artifact
+		generatedSha := hashGenerator(artifactURL, readBody)
+
+		// Verify the artifact signing itself
+		pubkeyRingReader, err := os.Open(publicKey)
+		if err != nil {
+			log.Error("Error opening publickey: ", err)
+		}
+		sigkeyRingReader, err := os.Open(signature)
+		if err != nil {
+			log.Error("Error opening signature: ", err)
+		}
+
+		var keyRing openpgp.EntityList
+		if isArmorProtected(pubkeyRingReader) {
+			keyRing, err = openpgp.ReadArmoredKeyRing(pubkeyRingReader)
+		} else {
+			keyRing, err = openpgp.ReadKeyRing(pubkeyRingReader)
+		}
+
+		dataReader := bytes.NewReader(readBody)
+
+		if isArmorProtected(sigkeyRingReader) {
+			_, err = openpgp.CheckArmoredDetachedSignature(keyRing, dataReader, sigkeyRingReader)
+		} else {
+			_, err = openpgp.CheckDetachedSignature(keyRing, dataReader, sigkeyRingReader)
+		}
+		if err != nil {
+			log.Error("Signature Verification failed: ", err)
+			os.Exit(1)
+		}
+		log.Info("Signature validation passed")
+
+		// Generate a file name based off the artifact hash
+		rekorFile := generateRekorFile(generatedSha)
+
+		log.Info("Building rekor file : ", rekorFile)
+
+		// Construct rekor json file
+		// We need to approach this in two ways
+		// as the public key and signature could be either
+		// armored or binary
+		if isArmorProtected(sigkeyRingReader) || isArmorProtected(pubkeyRingReader) {
+			rekorArmorJSON := RekorArmorEntry{
+				URL:       artifactURL,
+				SHA:       generatedSha,
+				Signature: sig,
+				PublicKey: pub_key,
+			}
+			file, _ := json.MarshalIndent(rekorArmorJSON, "", " ")
```

Comment:
Any reason for using indent here?

Reply:
No idea, maybe my IDE formatted it, I will change it